src/consensus/block_producer.rs

```diff
@@ -292,41 +311,50 @@ where
             } else {
                 duration_left
             };
-            let (payload, maybe_duration) =
+            let (payload, mut maybe_duration) =
                 produce_slice_payload(&self.txs_receiver, parent, time_for_slice).await;
+            if slice_index.is_first() {
+                let left = if let Some(duration) = maybe_duration {
+                    duration_left.saturating_sub(duration)
+                } else {
+                    duration_left.saturating_sub(self.delta_first_slice)
+                };
+                if left > Duration::ZERO {
+                    maybe_duration = Some(left);
+                } else {
+                    maybe_duration = None;
+                }
+            }
```

Comment:
```suggestion
            let (payload, maybe_duration) =
                produce_slice_payload(&self.txs_receiver, parent, time_for_slice).await;

            let maybe_duration = if slice_index.is_first() {
                let elapsed = maybe_duration.unwrap_or(self.delta_first_slice);
                duration_left.checked_sub(elapsed)
            } else {
                maybe_duration
            };

```

Reply:
is this more readable?

Reply:
we do, but I can also add an `assert` to `BlockProducer`